src/main/java/me/minidigger/hangar/controller/VersionsController.java

```diff
@@ -195,21 +295,34 @@ public Object downloadJar(@PathVariable Object author, @PathVariable Object slug
+    @PostMapping("/{author}/{slug}/versions/{version}/recommended")
+    public ModelAndView setRecommended(@PathVariable String author, @PathVariable String slug, @PathVariable String version) {
+        VersionData versionData = versionService.getVersionData(author, slug, version);
+        versionData.getP().getProject().setRecommendedVersionId(versionData.getV().getId());
+        projectDao.get().update(versionData.getP().getProject());
+        return new ModelAndView("redirect:" + routeHelper.getRouteUrl("versions.show", author, slug, version));
     }
 
+    @GlobalPermission(NamedPermission.REVIEWER)
     @Secured("ROLE_USER")
-    @RequestMapping("/{author}/{slug}/versions/{version}/restore")
-    public Object restore(@PathVariable Object author, @PathVariable Object slug, @PathVariable Object version) {
-        return null; // TODO implement restore request controller
+    @PostMapping(value = "/{author}/{slug}/versions/{version}/restore", consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    public ModelAndView restore(@PathVariable String author, @PathVariable String slug, @PathVariable String version, @RequestParam String comment, HttpServletRequest request) {
+        VersionData versionData = versionService.getVersionData(author, slug, version);
+        versionService.changeVisibility(versionData, Visibility.PUBLIC, comment, userService.getCurrentUser().getId());
+        userActionLogService.version(request, LoggedActionType.VERSION_DELETED.with(VersionContext.of(versionData.getP().getProject().getId(), versionData.getV().getId())), "Restore: " + comment, "");
+        return new ModelAndView("redirect:" + routeHelper.getRouteUrl("versions.show", author, slug, version));
     }
 
     @Secured("ROLE_USER")
-    @RequestMapping("/{author}/{slug}/versions/{version}/save")
-    public Object saveDescription(@PathVariable Object author, @PathVariable Object slug, @PathVariable Object version) {
-        return null; // TODO implement saveDescription request controller
+    @PostMapping(value = "/{author}/{slug}/versions/{version}/save", consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    public ModelAndView saveDescription(@PathVariable String author, @PathVariable String slug, @PathVariable String version, @RequestParam String content, HttpServletRequest request) {
+        VersionData versionData = versionService.getVersionData(author, slug, version);
+        String oldDesc = versionData.getV().getDescription();
+        String newDesc = content.trim();
+        versionData.getV().setDescription(newDesc);
+        versionService.update(versionData.getV());
+        userActionLogService.version(request, LoggedActionType.VERSION_DESCRIPTION_CHANGED.with(VersionContext.of(versionData.getP().getProject().getId(), versionData.getV().getId())), newDesc, oldDesc);
+        return new ModelAndView("redirect:" + routeHelper.getRouteUrl("versions.show", author, slug, version));
```

Comment:
You can also just return `return "redirect:" + routeHelper.getRouteUrl("versions.show", author, slug, version)` That would work because Spring will generate a `ModelAndView` out of it with all information.

Reply:
we should stay consistent here, most mappings that just generate some html return movelandview